tests/bootstrap.php

```diff
@@ -2,58 +2,109 @@
 }
 
-if ( ! file_exists( $_tests_dir . '/includes/functions.php' ) ) {
-	echo "Could not find {$_tests_dir}/includes/functions.php, have you run bin/install-wp-tests.sh ?" . PHP_EOL;
+// Forward custom PHPUnit Polyfills configuration to PHPUnit bootstrap file.
+$_phpunit_polyfills_path = getenv( 'WP_TESTS_PHPUNIT_POLYFILLS_PATH' );
+if ( false !== $_phpunit_polyfills_path ) {
+	define( 'WP_TESTS_PHPUNIT_POLYFILLS_PATH', $_phpunit_polyfills_path );
+}
+
+if ( ! file_exists( "{$_tests_dir}/includes/functions.php" ) ) {
+	echo "Could not find {$_tests_dir}/includes/functions.php, have you run bin/install-wp-tests.sh ?" . PHP_EOL; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
 	exit( 1 );
 }
 
 // Give access to tests_add_filter() function.
 require_once "{$_tests_dir}/includes/functions.php";
 
+/**
+ * Manually load the plugin or theme being tested.
+ */
+function _manually_load_plugin_or_theme() {
+	$entrypoint = _wordpressutils_find_entrypoint_file();
+    if ($entrypoint === true) {
+        _register_theme();
+    } elseif (is_string($entrypoint)) {
+        require $entrypoint;
+    }
+}
+
+/**
+ * Attempts to find the entrypoint of the theme or plugin being tested.
+ * Themes should have a functions.php file entrypoint.
+ * Plugins should have a <plugin-name>.php file entrypoint.
+ *
+ * @return string|bool Return the path to the <plugin-name>.php entrypoint for
+ *                     plugins. Return true for themes.
+ * @throws Exception If no single entrypoint php file can be found in the root directory.
+ */
+function _wordpressutils_find_entrypoint_file() {
+    $path = dirname( dirname( __FILE__ ), 4 );
+    $file_to_require = null;
+
+    // Get all php files in the plugin/theme root.
+    $files = glob($path . '/*.php');
+    $filenames = [];
+
+    // Get just the filename for each file found.
+    foreach($files as $file) {
+        $filenames[] = pathinfo($file, PATHINFO_FILENAME);
+    }
+
+    // If functions.php exists (theme), require it.
+    $key = array_search('functions', $filenames);
+    if ($key !== false) {
+        return true;
+    } else {
+        // Plugins are named <plugin-name>.php, attempt to filter php files to find this file.
+        $IGNORED_FILENAMES = ['index', 'uninstall'];
+        $filtered_filenames = array_diff($filenames, $IGNORED_FILENAMES);
+        $files_found = count($filtered_filenames);
+        if ($files_found === 1) {
+            $file_to_require = $files[array_key_first($filtered_filenames)];
+        } elseif ($files_found < 1) {
+            throw new Exception('No entrypoint php file found. Plugins should have a <plugin-name>.php and themes should have a functions.php.');
+        } else {
+            throw new Exception('Multiple potential entrypoint php files found. List of allowed *.php files in project root: <plugin-name>, ' . join($IGNORED_FILENAMES, ', '));
+        }
+    }
+
+    return $file_to_require;
+}
+
```

Comment:
Maybe this instead?

function _wordpressutils_find_entrypoint_file() {
    //Determine if we're in a theme or plugin
    $is_theme = false;
    $is_plugin = false;
    
    if (function_exists('get_template_directory')) {
        $template_dir = get_template_directory();
        $current_dir = dirname(dirname(__FILE__, 4));
        $is_theme = (strpos($current_dir, $template_dir) === 0);
    }
    
    if (!$is_theme && function_exists('plugin_basename')) {
        $plugin_dir = WP_PLUGIN_DIR;
        $current_dir = dirname(dirname(__FILE__, 4));
        $is_plugin = (strpos($current_dir, $plugin_dir) === 0);
    }

    // Get all PHP files in the root
    $path = dirname(dirname(__FILE__, 4));
    $files = glob($path . '/*.php');
    
    if ($is_theme) {
        // For themes, specifically look for functions.php
        $functions_file = $path . '/functions.php';
        return file_exists($functions_file) ? true : null;
    } elseif ($is_plugin) {
        // For plugins, look for the main plugin file
        $plugin_data = [];
        foreach ($files as $file) {
            // Skip known non-plugin files
            if (in_array(basename($file), ['index.php', 'uninstall.php'])) {
                continue;
            }
            
            // Look for the plugin header
            $default_headers = [
                'Plugin Name' => 'Plugin Name',
                'Version' => 'Version',
            ];
            $file_data = get_file_data($file, $default_headers);
            
            if (!empty($file_data['Plugin Name'])) {
                return $file;
            }
        }
    }
    
    throw new Exception('Could not determine if this is a theme or plugin, or could not find the entry point file.');
}

Reply:
Yeah this is better. I'm using get_file_data() to detect the plugin entrypoint now. I didn't think I had access to the WP functions in this context so I wasn't using them.